### PR TITLE
Phase 6: Geosolver — Geometry-Aware Feature Encoding for Airfoil Meshes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Explicit geometric feature encoding dramatically improves prediction accuracy on airfoil surfaces.** Geosolver (https://arxiv.org/pdf/2512.20399) shows that encoding geometric properties — surface curvature, wall-normal directions, and local element quality — as additional input features enables the network to understand the physics-geometry coupling that drives boundary layer behavior.

Our current model receives coordinates (x, y) and DSDF features but has no explicit representation of:
- **Surface curvature** — high curvature regions (leading edge, trailing edge) have the hardest-to-predict pressure distributions
- **Wall-normal direction** — the boundary layer structure depends on the normal to the nearest surface
- **Surface proximity** — nodes very close to the surface need different treatment than far-field nodes

By adding these geometric features, we give the Transolver direct access to information it currently must infer from raw coordinates. This is especially relevant for our worst metric `p_tan = 30.01` (tandem foil pressure), where the complex geometry of two foils creates intricate wake interactions.

Paper: https://arxiv.org/pdf/2512.20399

### Step 1: Compute Geometric Features in Preprocessing

Add a function that computes geometric features from mesh coordinates and surface masks. These are computed ONCE per sample during data loading (no training overhead):

```python
def compute_geo_features(coords, surf_mask):
    """Compute geometric features for each mesh node.
    
    Args:
        coords: [N, 2] node coordinates (x, y)
        surf_mask: [N] boolean surface node mask
    
    Returns:
        geo_feats: [N, 4] geometric features:
            - wall_dist: distance to nearest surface node (log-scaled)
            - curvature: local surface curvature (0 for volume nodes)
            - normal_x, normal_y: wall-normal direction
    """
    surf_coords = coords[surf_mask]  # [S, 2]
    
    # Wall distance for all nodes
    dists = torch.cdist(coords.unsqueeze(0), surf_coords.unsqueeze(0)).squeeze(0)  # [N, S]
    wall_dist, nearest_idx = dists.min(dim=1)  # [N]
    wall_dist_log = torch.log1p(wall_dist)  # log-scale for better distribution
    
    # Wall-normal direction (vector from nearest surface point to this node)
    nearest_surf = surf_coords[nearest_idx]  # [N, 2]
    normal_vec = coords - nearest_surf  # [N, 2]
    normal_norm = normal_vec.norm(dim=1, keepdim=True).clamp(min=1e-6)
    normal_vec = normal_vec / normal_norm  # unit normal
    # For surface nodes themselves, use local tangent rotated 90°
    
    # Surface curvature (discrete curvature at surface nodes, 0 elsewhere)
    curvature = torch.zeros(coords.shape[0])
    if surf_mask.sum() > 2:
        # Sort surface nodes by angle from centroid for ordered traversal
        centroid = surf_coords.mean(dim=0)
        angles = torch.atan2(surf_coords[:, 1] - centroid[1], 
                            surf_coords[:, 0] - centroid[0])
        order = angles.argsort()
        ordered = surf_coords[order]
        # Discrete curvature: angle change between consecutive segments
        tangents = ordered.roll(-1, 0) - ordered
        tangent_angles = torch.atan2(tangents[:, 1], tangents[:, 0])
        curv = (tangent_angles.roll(-1, 0) - tangent_angles)
        curv = torch.atan2(torch.sin(curv), torch.cos(curv))  # wrap to [-π, π]
        seg_len = tangents.norm(dim=1).clamp(min=1e-6)
        curv = curv / seg_len  # curvature = dθ/ds
        # Map back to original surface ordering
        inv_order = torch.empty_like(order)
        inv_order[order] = torch.arange(len(order))
        surf_indices = torch.where(surf_mask)[0]
        curvature[surf_indices] = curv[inv_order]
    
    geo_feats = torch.stack([wall_dist_log, curvature, normal_vec[:, 0], normal_vec[:, 1]], dim=1)
    return geo_feats
```

### Step 2: Integrate into Model Input

Add a `--geo_features` flag. When enabled:
1. Compute geo features in the data loading / collate step (so it's cached, not recomputed each epoch)
2. Concatenate 4 extra features to the input `x` tensor, so `fun_dim` increases by 4
3. Update the `GatedMLP2` preprocessing input dimension accordingly: `fun_dim + space_dim + 4`

**Important:** The curvature computation uses `torch.cdist` which can be expensive for large meshes. Pre-compute and cache if needed. Our meshes are ~25K nodes so `cdist` with ~2K surface nodes should be fine.

### Step 3: Geometry-Conditioned Attention (Advanced)

Add a `--geo_attn` flag. When enabled, use wall distance as an additional spatial bias in the slice attention:
```python
# In Physics_Attention_Irregular_Mesh.forward():
# geo_bias: [B, N, 1] wall distance (log-scaled)
if geo_bias is not None:
    # Nodes at similar wall distances should attend to each other more
    geo_slice_bias = self.geo_proj(geo_bias)  # [B, N, slice_num]
    slice_logits = slice_logits + 0.1 * geo_slice_bias.unsqueeze(1)
```

Add `self.geo_proj = nn.Linear(1, slice_num)` with zero-init to `Physics_Attention_Irregular_Mesh.__init__`.

### Experiment Sweep (8 GPUs)

| GPU | Config | Flags |
|-----|--------|-------|
| 0 | Baseline | `--seed 42` |
| 1 | Baseline | `--seed 43` |
| 2 | Geo features only | `--geo_features --seed 42` |
| 3 | Geo features only | `--geo_features --seed 43` |
| 4 | Geo features + geo attention | `--geo_features --geo_attn --seed 42` |
| 5 | Geo features + geo attention | `--geo_features --geo_attn --seed 43` |
| 6 | Geo features + geo attn + higher surf_weight | `--geo_features --geo_attn --surf_weight 30 --seed 42` |
| 7 | Geo features + geo attn + higher surf_weight | `--geo_features --geo_attn --surf_weight 30 --seed 43` |

**Base command for all runs:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse \
  --wandb_group "phase6/geosolver" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep --residual_prediction \
  --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  [GPU-specific flags above]
```

### Important Notes
- The curvature calculation assumes surface nodes can be ordered. For our tandem foil meshes with TWO airfoils, you need to handle each foil separately (use connected components or the tandem mask to split surface nodes into two groups before computing curvature)
- Wall distance should be computed from ALL surface nodes (both foils), not just the nearest foil
- Pre-compute geo features in `__getitem__` and cache them, not in the forward pass
- The `torch.cdist` call is O(N * S) — with N=25K and S=2K this is ~50M operations, fast on GPU

## Baseline

| Metric | Mean | Std | Best single |
|--------|------|-----|-------------|
| val/loss | 0.383 | 0.003 | 0.378 |
| p_in | 12.95 | 0.30 | 12.1 |
| p_oodc | 8.31 | 0.18 | 8.1 |
| p_tan | 30.01 | 0.52 | 29.2 |
| p_re | 6.70 | 0.10 | 6.5 |

W&B group: `phase5/surface-refine-8seed`

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep --residual_prediction \
  --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2009 because ADVISOR: Closing — the researcher-agent found that GeoTransolver/Geosolver was already tested in PR #1989 and failed decisively (all variants 9.8% worse than baseline, learned gate values near zero meaning the model explicitly rejected the geometry context, 20% throughput loss, +5GB VRAM, multiple torch.compile crashes). The 24-dim input feature vector already encodes geometry info (dsdf, saf, curvature, coordinates). Reassigning you to NOBLE instead — a much more promising direction..
